include/ProxyWifi/ProxyWifiService.hpp

```diff
@@ -95,30 +87,86 @@ struct OnDisconnectionArgs {
     DOT11_SSID disconnectedNetwork{};
 };
 
-struct ProxyWifiCallbacks
+enum class OperationType
 {
-    /// @brief Callback executed on a connection event
-    /// If the `EventSource` is the `Host`, it means indicated host interface connected
-    /// If the `EventSource` is the `Guest`, it means the guest is connecting (the connection request will be
-    /// completed only after this callback returns)
-    std::function<void(EventSource, const OnConnectionArgs&)> OnConnection;
-
-    /// @brief Callback executed on a disconnection event
-    /// If the `EventSource` is the `Host`, it means indicated host interface disconnected
-    /// If the `EventSource` is the `Guest`, it means the guest is disconnecting (the disconnection request will be
-    /// completed only after this callback returns)
-    std::function<void(EventSource, const OnDisconnectionArgs&)> OnDisconnection;
-
-    /// @brief Callback executed on the guest connection request that will cause a host interface to connect
-    /// The callback is invoked when before interfaces are connected/disconnected on the host, and after the connection
-    /// attempt is completed, whether it succeeds or fails.
-    std::function<void(GuestConnectStatus)> OnGuestConnectRequestProgress;
-
-    /// @brief Callback providing a list of networks that will be simulated by the Wi-Fi proxy
-    /// They will be shown as open networks, and are considered as permanently connected for the purpose of notifications
-    std::function<std::vector<WifiNetworkInfo>()> ProvideFakeNetworks;
+    GuestDirected,
+    HostMirroring
 };
 
+// Temporarily disable the "unused argument warning"
+#pragma warning(push)
+#pragma warning(disable : 4100)
+class ProxyWifiObserver
+{
+public:
+    struct ConnectCompleteArgs {
+        GUID interfaceGuid;
+        DOT11_SSID ssid;
+        DOT11_AUTH_ALGORITHM authAlgo;
+    };
+    struct DisconnectCompleteArgs {
+        GUID interfaceGuid;
+        DOT11_SSID ssid;
+    };
+
+    /// @brief An host WiFi interface connected to a network
+    virtual void OnHostConnection(const ConnectCompleteArgs& connectionInfo) noexcept
+    {
+    }
+
+    /// @brief An host WiFi interface disconnected from a network
+    virtual void OnHostDisconnection(const DisconnectCompleteArgs& disconnectionInfo) noexcept
+    {
+    }
+
+    /// @brief The guest requested a connection to a network
+    /// If `type == OperationType::HostMirroring`, an host inteface is already connected to the network, otherwise, one will be
+    /// connected The connection won't proceed until the callback returns
+    virtual void OnGuestConnectionRequest(OperationType type, const DOT11_SSID& ssid) noexcept
```

Comment:
Should `ssid` maybe be wrapped in a structure like the args for other callbacks?

```suggestion
    struct GuestConnectionRequestArgs {
        DOT11_SSID ssid;
    };
    virtual void OnGuestConnectionRequest(OperationType type, const GuestConnectionRequestArgs& connectionRequestArgs) noexcept
```

Would make this more consistent and protect against any possible future additions/changes to those args. Same for `OnGuestDisconnectionRequest` below.

Reply:
Yeah, you might be right. I was trying to find a balance between packaging and not creating struct with a single value, but maybe this is just being lazy.

Reply:
Not lazy :)